File: hw_10244302412_fulladder.py
def half_adder(a,b):
    sum=a^b
    carry=a&b
    return sum,carry
def full_adder(a,b,c):
    sum0,carry0=half_adder(a,b)
    sum1,carry1=half_adder(sum0,c)
    sum2,carry2=sum1,carry0|carry1
    return sum2,carry2
def two_adder(A0,A1,B0,B1):
    sum0,carry0=half_adder(A0,B0)
    sum1,carry1=full_adder(A1,B1,carry0)
    return carry1*100+sum1*10+sum0
result1=two_adder(1,1,1,1);
print("11+11=",result1)
result2=two_adder(1,0,0,1);
print("01+10=",result2)

File: works/hw_10244302412_fulladder.c
#include<stdio.h>
typedef struct{
    int sum;
    int carry;
} BIT;
BIT half_adder(int a,int b){
    return (BIT){a^b,a&b};
}
BIT full_adder(int a,int b,int c){
    BIT v1=half_adder(a,b);
    BIT v2=half_adder(v1.sum,c);
    return (BIT){v2.sum,v1.carry|v2.carry};
}
int two_adder(int A0,int A1,int B0,int B1){
    BIT v0=half_adder(A0,B0);
    BIT v1=full_adder(A1,B1,v0.carry);
    return v1.carry*100+v1.sum*10+v0.sum;
}
int main(){
    printf("11+11=%d\n",two_adder(1,1,1,1));
    printf("10+01=%d\n",two_adder(1,0,0,1));
    return 0;
}
